### PR TITLE
[Yaml] Fix wrong line number when comments are inserted in the middle of a block.

### PR DESCRIPTION
**Pull request by paradajozsef**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15437 |
| License | MIT |
| Doc PR | - |

@xabbuh what is your opinion a solution like this? (counting the skipped comment lines and when exception occurs add them to the current line count.)

### CODE REVIEW

**Comment by xabbuh** [comment]
@paradajozsef I think the fix is a bit more complex. For example, I do not think that this will solve the issue for the following example too:

```
foo:
    -
        # foobar
        baz: 123
bar:
    -
        # bar
        bar: "123",
```

Can you add it as a test case too?

**Comment by paradajozsef** [comment]
@xabbuh Sorry for late answer. I added the test case and as always you are right again. :smiley: In this case the line number is 9, but it should be 8. :confused: Still needs work.

**Comment by paradajozsef** [comment]
@xabbuh I changed the fix. `getNextEmbedBlock` will only skip comment lines, if comments are not in a collection. WDUT?

The above test case passes, failing VarDumper tests unrelated imo.

**Comment by paradajozsef** [comment]
Status: Needs Review

File: src/Symfony/Component/Yaml/Parser.php [hunk 1]
```diff
@@ -769,4 +777,44 @@ private function isBlockScalarHeader()
...
+
+        return '' !== $ltrimmedLine && '-' === $ltrimmedLine[0];
+    }
+
+    /**
+     * Tests whether or not the current comment line is in a collection.
+     *
+     * @return bool
+     */
+    private function isPreviousNonCommentLineIsCollectionItem()
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: method name is weird. It contains ``is`` twice

File: src/Symfony/Component/Yaml/Parser.php [hunk 2]
```diff
@@ -769,4 +777,44 @@ private function isBlockScalarHeader()
...
+     */
+    private function isCurrentLineCollectionItem()
+    {
+        $ltrimmedLine = ltrim($this->currentLine, ' ');
+
+        return '' !== $ltrimmedLine && '-' === $ltrimmedLine[0];
+    }
+
+    /**
+     * Tests whether or not the current comment line is in a collection.
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: ``or not`` is useless here

**Comment by fabpot** [comment]
ping @xabbuh

**Comment by paradajozsef** [comment]
I apologize to everyone for late answer. (I've been lack of internet at home lately because of moving to another place.)

@stof Thanks for the review. I changed the comment and method name. What about `checkIfPreviousNonCommentLineIsCollectionItem`? It's still kind of wierd imo. Any other ide is very welcome! Thanks!

**Comment by xabbuh** [comment]
@paradajozsef Can you please rebase and push again? Looks like Travis otherwise will not run builds for your changes otherwise.

**Comment by paradajozsef** [comment]
Sure, I can do it tonight.

**Comment by stof** [comment]
@xabbuh the reason why Travis does not run tests is because the branch conflicts. As Travis is actually running PR tests on a merge commit rather than on the feature branch, it cannot run them in case of conflicts

**Comment by paradajozsef** [comment]
Rebase is done, tests are green.

File: src/Symfony/Component/Yaml/Parser.php [hunk 3]
```diff
@@ -804,4 +812,44 @@ private function isBlockScalarHeader()
...
+    /**
+     * Tests whether the current comment line is in a collection.
+     *
+     * @return bool
+     */
+    private function checkIfPreviousNonCommentLineIsCollectionItem()
+    {
+        $isCollectionItem = false;
+        $moves = 0;
+        while($this->moveToPreviousLine()) {
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: please add a space before the parenthesis

File: src/Symfony/Component/Yaml/Parser.php [hunk 4]
```diff
@@ -804,4 +812,44 @@ private function isBlockScalarHeader()
...
+            // If previous line is a comment, move back again.
+            if ($this->isCurrentLineComment()) {
+                continue;
+            }
+            $isCollectionItem = $this->isCurrentLineCollectionItem();
+            break;
+        }
+
+        // Move parser back to previous line.
+        while($moves > 0) {
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: same here

File: src/Symfony/Component/Yaml/Parser.php [hunk 5]
```diff
@@ -804,4 +812,44 @@ private function isBlockScalarHeader()
     {
         return (bool) preg_match('~'.self::BLOCK_SCALAR_HEADER_PATTERN.'$~', $this->currentLine);
     }
+
+    /**
+     * Returns true if the current line is a collection item.
+     *
+     * @return bool Returns true if the current line is a collection item line, false otherwise.
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: The description of the returnt ype here is not needed (it's already clear from the method's description), just use `@return true`.

**Comment by xabbuh** [comment]
👍 

Status: Reviewed

ping @symfony/deciders

**Comment by paradajozsef** [comment]
@xabbuh Thx for the review, I fixed them.

**Comment by fabpot** [comment]
Thank you @paradajozsef.

**Pull request closed without merging**